File: avail-js/docs/book/src/examples/validator.ts
import { SDK, BN, utils } from "./../../../../src/index"

export async function run() {
  const sdk = await SDK.New(SDK.localEndpoint())
  const api = sdk.api

  const account = SDK.alice()

  // Bond minValidatorBond or 1 AVAIL token
  const minValidatorBond: BN = ((await api.query.staking.minValidatorBond()) as any) || SDK.oneAvail()

  // Bond
  const bondTx = sdk.tx.staking.bond(minValidatorBond, "Staked")
  const _r1 = (await bondTx.executeWaitForInclusion(account))._unsafeUnwrap()

  // Generate Session Keys
  const keysBytes = await api.rpc.author.rotateKeys()
  const keys = utils.deconstruct_session_keys(keysBytes.toString())

  // Set Keys
  const setKeysTx = sdk.tx.session.setKeys(keys)
  const _r2 = (await setKeysTx.executeWaitForInclusion(account))._unsafeUnwrap()

  // Validate
  const validateTx = sdk.tx.staking.validate(50, false)
  const _r3 = (await validateTx.executeWaitForInclusion(account))._unsafeUnwrap()
}

File: avail-rust/docs/book/src/examples/validator.rs
use avail_rust::{
	avail, error::ClientError, transactions::staking::Commission, utils, Options,
	RewardDestination, SDK,
};

pub async fn run() -> Result<(), ClientError> {
	let sdk = SDK::new(SDK::local_endpoint()).await?;

	let account = SDK::alice()?;

	// Bond min_validator_bond or 1 AVAIL token
	let storage = sdk.online_client.storage().at_latest().await?;
	let min_validator_bond = storage
		.fetch(&avail::storage().staking().min_validator_bond())
		.await?
		.unwrap_or_else(|| SDK::one_avail());

	let payee = RewardDestination::Staked;

	// Bond
	let tx = sdk.tx.staking.bond(min_validator_bond, payee);
	let options = Some(Options::new().nonce(avail_rust::Nonce::BestBlockAndTxPool));
	_ = tx.execute_wait_for_inclusion(&account, options).await?;

	// Generate Session Keys
	let keys = sdk.rpc.author.rotate_keys().await?;
	let keys = utils::deconstruct_session_keys(keys)?;

	// Set Keys
	let tx = sdk.tx.session.set_keys(keys);
	_ = tx.execute_wait_for_inclusion(&account, options).await?;

	// Validate
	let commission = Commission::new(10)?;
	let tx = sdk.tx.staking.validate(commission, false);
	_ = tx.execute_wait_for_inclusion(&account, options).await?;

	Ok(())
}
